### PR TITLE
[16.0][ADD] sale_stock_restocking_fee_invoicing

### PR DESCRIPTION
**Pull request by glitchov**
This module extends the functionality of the stock picking return wizard
to allow you to charge restocking fee in case of return of goods by a
customer.

### CODE REVIEW

File: sale_stock_restocking_fee_invoicing/README.rst [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+* or *Inventory* - *Configuration* - *Settings* - *Valuation*
+
+Once there you can change the item Restocking Fee Product.
+
+Usage
+=====
+
+Once installed, A check box is added on the stock return wizard to specify if
+restocking fee must be charged. The same field is added on each line into
+the wizard. You are therefore able to choice to charge for all the goods
```
Comment (jbaudoux) [inline sale_stock_restocking_fee_invoicing/README.rst:53]:
```suggestion
the wizard. You are therefore able to choose to charge for all the goods
```

File: sale_stock_restocking_fee_invoicing/models/res_company.py [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+        help="It is possible to specify on a customer whether restocking fee "
+        "must be charged for returning goods. In the case where fee are to "
+        "be applied, the product specified here is added to the sale"
+        " order to charge these restocking fee.",
+    )
+
+    @api.model
+    def _default_restocking_fee_product_id(self):
+        product = self.env.ref(
+            "sale_stock_restocking_fee_invoicing." "product_restocking_fee",
```
Comment (jbaudoux) [inline sale_stock_restocking_fee_invoicing/models/res_company.py:26]:
```suggestion
            "sale_stock_restocking_fee_invoicing.product_restocking_fee",
```

File: sale_stock_restocking_fee_invoicing/models/sale_order.py [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+from odoo.exceptions import UserError
+from odoo.fields import Command
+
+
+class SaleOrder(models.Model):
+
+    _inherit = "sale.order"
+
+    def _get_restocking_fee_line_name(self, stock_move):
+        """Return the de name for the so line. This method should be called
```
Comment (jbaudoux) [inline sale_stock_restocking_fee_invoicing/models/sale_order.py:14]:
```suggestion
        """Return the SO line description for the restocking fee. This method should be called
```

File: sale_stock_restocking_fee_invoicing/models/stock_picking.py [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    _inherit = "stock.picking"
+
+    is_customer_return = fields.Boolean(compute="_compute_is_customer_return")
+
+    is_customer_out = fields.Boolean(compute="_compute_is_customer_out")
+
+    @api.depends("location_id")
+    def _compute_is_customer_return(self):
+        for rec in self:
+            rec.is_customer_return = rec.location_id.usage == "customer"
```
Comment (jbaudoux) [inline sale_stock_restocking_fee_invoicing/models/stock_picking.py:18]: To be safer
```suggestion
            rec.is_customer_return = rec.location_id.usage == "customer" and rec.location_dest_id.usage != "customer" 
```

File: sale_stock_restocking_fee_invoicing/models/stock_picking.py [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+    @api.depends("location_dest_id")
+    def _compute_is_customer_out(self):
+        for rec in self:
+            rec.is_customer_out = rec.location_dest_id.usage == "customer"
```
Comment (jbaudoux) [inline sale_stock_restocking_fee_invoicing/models/stock_picking.py:23]: To be safer
```suggestion
            rec.is_customer_out = rec.location_id.usage != "customer" and rec.location_dest_id.usage == "customer"
```

File: sale_stock_restocking_fee_invoicing/wizards/stock_return_picking.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+        if not (
+            "charge_restocking_fee" in fields
+            or "product_return_moves" in fields
+            or "is_customer_return" in fields
+        ):
+            return res
+        picking = self.env["stock.picking"].browse(self.env.context["active_id"])
+        charge_restocking_fee = picking.partner_id.charge_restocking_fee
+        if "is_customer_return" in fields:
+            res["is_customer_return"] = picking.location_dest_id.usage == "customer"
```
Comment (jbaudoux) [inline sale_stock_restocking_fee_invoicing/wizards/stock_return_picking.py:31]: Use the computed field you created
```suggestion
            res["is_customer_return"] = picking.is_customer_out
```

**Review by jbaudoux — APPROVED** [review]
Some minor non blocking remarks

File: sale_stock_restocking_fee_invoicing/__manifest__.py [hunk 7]
```diff
@@ -0,0 +1,24 @@
...
+
+{
+    "name": "Sale Stock Restocking Fee Invoicing",
+    "summary": """
+        On demand charge restocking fee for accepting returned goods .""",
+    "version": "16.0.1.0.0",
+    "license": "AGPL-3",
+    "author": "ACSONE SA/NV,Odoo Community Association (OCA)",
+    "website": "https://github.com/OCA/account-invoicing",
+    "depends": ["sale", "sale_stock", "stock", "stock_account"],
```
Comment (jbaudoux) [inline sale_stock_restocking_fee_invoicing/__manifest__.py:12]: Why is stock_account required?

**Review by jbaudoux — CHANGES_REQUESTED** [review]
In my opinion, this PR is not in the right repository. Besides the module name containing invoicing because the added SO line is meant to be later invoiced by the standard process, I don't see anything directly related to invoicing.
I would move this PR to `stock-logistics-workflow` as the return of goods is a logistics process.

File: sale_stock_restocking_fee_invoicing/__manifest__.py (hunk 7 above)
Comment (glitchov) [reply sale_stock_restocking_fee_invoicing/__manifest__.py:12]: To put the settings inside the valuation_setting_container which comes with stock_account. Alternatively i can put them in the warehouse_setting_container to remove this dependency but sale_stock already depends on it.

**Comment by glitchov** [comment]
Transferred to https://github.com/OCA/stock-logistics-workflow/pull/1251

**Pull request closed without merging**